packages/redis_client/test/src/redis_client_test.dart

```diff
@@ -1,10 +1,157 @@
+import 'dart:io';
+
 import 'package:redis_client/redis_client.dart';
 import 'package:test/test.dart';
 
 void main() {
   group(RedisClient, () {
-    test('can be instantiated', () {
-      expect(const RedisClient(), isNotNull);
+    late RedisClient client;
+
+    setUp(() async {
+      client = RedisClient();
+      await client.connect();
+    });
+
+    tearDown(() async {
+      try {
+        await client.sendCommand(['RESET']);
+        await client.sendCommand(['FLUSHALL']);
+        await client.close();
+      } catch (_) {
+        // ignore
+      }
+    });
+
+    test('throws SocketException when connection times out', () async {
+      final client = RedisClient(
+        socket: const RedisSocketOptions(timeout: Duration(microseconds: 1)),
+      );
+      await expectLater(
+        () => client.connect(maxConnectionAttempts: 1),
+        throwsA(
+          isA<SocketException>().having(
+            (e) => e.message,
+            'message',
+            contains('Connection timed out'),
+          ),
+        ),
+      );
+    });
+
+    test('throws SocketException after max connection attempts', () async {
+      final client = RedisClient(
+        socket: const RedisSocketOptions(port: 1234),
+      );
+      await expectLater(
+        () => client.connect(maxConnectionAttempts: 1),
+        throwsA(
+          isA<SocketException>().having(
+            (e) => e.message,
+            'message',
+            contains('Connection refused'),
+          ),
+        ),
+      );
+    });
+
+    test('throws SocketException after disconnect', () async {
+      final client = RedisClient(
+        socket: const RedisSocketOptions(port: 1234),
+      );
+      await expectLater(
+        () => client.connect(maxConnectionAttempts: 0),
+        throwsA(
+          isA<SocketException>().having(
+            (e) => '$e',
+            'message',
+            contains('Connection retry limit exceeded'),
+          ),
+        ),
+      );
+    });
+
+    test('throws StateError when closed', () async {
+      await client.close();
+      await expectLater(
+        client.connect(),
+        throwsA(
+          isA<StateError>().having(
+            (e) => e.message,
+            'message',
+            'RedisClient has been closed.',
+          ),
+        ),
+      );
+    });
+
+    group('AUTH', () {
+      test('is required', () async {
+        await expectLater(
+          client.get(key: 'foo'),
+          throwsA(
+            isA<StateError>().having(
+              (e) => e.message,
+              'message',
+              contains('-NOAUTH Authentication required.'),
+            ),
+          ),
+        );
+      });
+
+      test('fails when username is incorrect', () async {
+        await expectLater(
+          client.auth(username: 'shorebird', password: 'password'),
+          completion(equals(false)),
+        );
```

Comment:
Stylistic question: is there an advantage to this form of `expect` over `expect(await myFunction(), isTrue)`?

Reply:
Either way, `isTrue`/`isFalse` is probably more idiomatic use of the matcher framework

Reply:
> Stylistic question: is there an advantage to this form of `expect` over `expect(await myFunction(), isTrue)`?

I believe the matcher description will be nicer if the test fails with the former as opposed to the later but it's mostly habitual 😆